docs(perf): add RAG performance measurement methodology (#490)
NVIDIA-AI-Blueprints/rag: docs/perf-benchmarks.md

<!--
  SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  SPDX-License-Identifier: Apache-2.0
-->

# RAG Performance Measurement Methodology

> [GenAI Perf](https://github.com/triton-inference-server/perf_analyzer/tree/main/genai-perf), NVIDIA's open-source benchmarking tool, is employed to evaluate the end-to-end performance of the RAG pipeline under realistic load conditions. By sweeping across multiple concurrency levels, it provides a reproducible and consistent basis for comparing different RAG deployment configurations.

Reviewer: ```suggestion
[GenAI Perf](https://github.com/triton-inference-server/perf_analyzer/tree/main/genai-perf), NVIDIA’s open-source benchmarking tool evaluates end-to-end RAG pipeline performance under realistic load conditions. By testing across varying levels of concurrency, it offers a consistent and reproducible way to compare different RAG deployment configurations.
```

## Key Terms

| Term | Description |
|------|-------------|
| **Total Requests** | The total number of questions issued to the RAG server in a single benchmark run. Controls the size of the workload and is kept constant across configurations for fair comparison. |
| **Concurrency** | The number of simultaneously active worker threads sending requests to the server. A higher concurrency simulates a heavier multi-user load. |

> | **N_Times** | The number of measured benchmark iterations performed after warm-up. Multiple iterations improve statistical stability of the reported metrics. In all experiments reported in this document, N_Times is set to 3. |

Reviewer: ```suggestion
| **N_Times** | The number of measured benchmark iterations performed after warm-up. Multiple iterations improve statistical stability of the reported metrics. In all experiments reported in this document, `N_Times` is set to 3. |
```

| **Input Sequence Length (ISL)** | The number of tokens in the input prompt sent to the RAG server. |
| **Output Sequence Length (OSL)** | The number of tokens in the RAG server's generated response. |
| **TTFT (Time to First Token)** | The elapsed time from when a request is submitted until the first output token is returned. A key indicator of perceived responsiveness in streaming deployments. |
| **Inter-Token Latency (ITL)** | Defined as (E2E Latency - TTFT) / (OSL - 1), where OSL is the number of output tokens generated per request, averaged across all requests in the benchmark run. |
| **KV Cache** | A memory buffer on the GPU (HBM) storing the key-value attention states computed for all tokens in an active request. KV cache size grows with sequence length, number of model layers, and hidden dimension. When aggregate KV cache across concurrent requests saturates HBM, new requests queue rather than execute, driving up TTFT. |
| **HBM (High Bandwidth Memory)** | The GPU's on-chip memory (e.g., 80 GB on H100). |

> | **Prefill** | The phase where the model processes the entire input prompt in parallel to build the KV cache. |

Reviewer: ```suggestion
| **Prefill** | The stage in which the model processes the full input prompt simultaneously to construct the KV cache. |
```

|---------|--------------------:|------------------|------|-------------------|
| [RagBattlePacket](https://www.eyelevel.ai/post/most-accurate-rag) | 92 | Deloitte public tax PDFs | 1,146 pages | 92 questions across text, tabular, and graphical categories; visually dense corpus with rich tables and figures requiring cross-modal understanding |
| [KG-RAG](https://github.com/docugami/KG-RAG-datasets/tree/main/sec-10-q/data/v1) | 195 | SEC 10-Q PDFs + KG triples | 1,037 pages | Entity-centric factual QA over structured financial filings; questions target specific named entities and numerical facts with minimal visual content |
| [HotPotQA](https://huggingface.co/datasets/hotpotqa/hotpot_qa) | 979 | Wikipedia paragraphs | ~113K QA pairs; 2,673 source documents | Multi-hop reasoning requiring the model to chain facts across multiple documents (bridge and comparison); plain text only with no tables or figures |
| [BO767](https://digitalcorpora.org/) | 487 | 767 PDFs | 54,730 pages | Varied, heterogeneous content across a large-scale mixed corpus of forensic and operational documents; high proportion of image and structured content alongside text |

## How It Works

> 

Reviewer: ```suggestion
The following sections describe how benchmarking works.
```

### Sequential Request Dispatch

Each thread draws questions from the pool one at a time in sequence, taking the next request only after a response to the current one has been received. This models realistic per-user session behavior and avoids artificially inflating throughput through intra-thread batching.

### Warm-Up and Measured Runs

Before recording any metrics, GenAI Perf executes an initial warm-up run to bring the RAG servers to a steady operational state, eliminating cold-start artifacts. The benchmark is then repeated for a configurable number of iterations (`N_Times`), ensuring that the collected statistics are stable and reproducible.

### Performance Result Collection

For every request across all threads and iterations, GenAI Perf records timing and outcome data. These are aggregated into a performance result store, from which key metrics — including Time-to-First-Token and Inter-Token Latency — are computed and reported per concurrency level.

## Configuration and Performance Results

> 

Reviewer: ```suggestion
The following sections provide further information about configuration and performance results.
```

### Configuration and Setup

The following deployment configurations are evaluated:

- **LLM-49B** — [Llama-3.3-Nemotron-Super-49B](https://build.nvidia.com/nvidia/llama-3_3-nemotron-super-49b-v1_5/modelcard): A key feature is the reasoning toggle — setting the system prompt to "detailed thinking on" causes the model to generate an internal chain-of-thought before the final answer; "detailed thinking off" produces a direct response. This is a system-prompt switch — no weight change occurs between modes.
- **VLM nano** — [Nemotron Nano VL](https://build.nvidia.com/nvidia/nemotron-nano-12b-v2-vl/modelcard): Its smaller parameter count means significantly lower KV cache memory consumption per request compared to the 49B model.
- **Ingestion setup:**

>   - *Default ingestion:* The default ingestion that RAG 2.4.0 uses.

Reviewer: ```suggestion
  - Default ingestion: The default ingestion that RAG 2.4.0 uses.
```

>   - *Default ingestion with VLM image captioning enabled:* Details in [Image Captioning Support](image_captioning.md). VLM is enabled during ingestion to extract image and structured content from documents, and also enabled at query time to process retrieved image and structured chunks via the multimodal pipeline.

Reviewer: ```suggestion
  - Default ingestion with VLM image captioning enabled: Details in [Image Captioning Support](image_captioning.md). VLM is enabled during ingestion to extract image and structured content from documents, and also enabled at query time to process retrieved image and structured chunks via the multimodal pipeline.
```

| # | LLM Model | Embedding Model | Reasoning On/Off | Ingestion Method |
|---|-----------|----------------|-------------------|-----------------|
| 1 | LLM-49B | Default Embedding | On | Default Ingestion |
| 2 | LLM-49B | Default Embedding | Off | Default Ingestion |
| 3 | VLM nano | Default Embedding | On | Default Ingestion with VLM image captioning enabled |
| 4 | VLM nano | Default Embedding | Off | Default Ingestion with VLM image captioning enabled |

### Impact Factors

> **TTFT (Time to First Token) at high concurrency is primarily determined by KV cache memory pressure.** Each active request occupies HBM proportional to sequence length × model layers × hidden dimension. When aggregate KV cache across concurrent requests saturates HBM, incoming requests queue, driving up p95 TTFT.

Reviewer: ```suggestion
TTFT (Time to First Token) at high concurrency is primarily determined by KV cache memory pressure. Each active request occupies HBM proportional to sequence length × model layers × hidden dimension. When aggregate KV cache across concurrent requests saturates HBM, incoming requests queue, driving up p95 TTFT.
```

- VLM-Reasoning-On records approximately 6× lower ITL than VLM-Reasoning-Off on this dataset. Both VLM modes go through the same image processing pipeline, so the delta is driven entirely by output length. VLM-Reasoning-Off produces short outputs — the 12B model's small KV cache already allows a large active batch, and short per-request lifetimes keep that batch continuously full, sustaining high decode contention. When reasoning is enabled, each request generates a long chain-of-thought over the visually complex tax documents before producing its final answer, holding each request in the active batch for a far longer duration, preventing the scheduler from admitting new requests and shrinking the effective batch size significantly. The magnitude of the drop is amplified by RagBattlePacket's document complexity, which elicits longer reasoning chains than simpler text corpora.

#### HotPotQA

![HotPotQA — H100 Performance](assets/perf-benchmarks/hotpotqa_h100_performance.png)

- LLM-Reasoning-On produces the highest p95 TTFT of all datasets, exceeding 250s at concurrency=125.

>   - *Reason:* HotPotQA's multi-hop questions require chaining facts across multiple source documents, which elicits longer reasoning chains from a thinking-enabled model. Each request holds a large KV cache slot (49B model) for an extended duration, accelerating HBM saturation at scale.

Reviewer: ```suggestion
  - Reason: HotPotQA's multi-hop questions require chaining facts across multiple source documents, which elicits longer reasoning chains from a thinking-enabled model. Each request holds a large KV cache slot (49B model) for an extended duration, accelerating HBM saturation at scale.
```

- VLM-Reasoning-Off records the lowest TTFT.

>   - *Reason:* HotPotQA source documents are plain Wikipedia text containing no tables or figures, so VLM ingestion does not inflate retrieved context. The model-size advantage of VLM nano translates directly into lower KV cache pressure and reduced queuing latency.

Reviewer: ```suggestion
  - Reason: HotPotQA source documents are plain Wikipedia text containing no tables or figures, so VLM ingestion does not inflate retrieved context. The model-size advantage of VLM nano translates directly into lower KV cache pressure and reduced queuing latency.
```
